app/src/org/commcare/mediadownload/MissingMediaDownloadHelper.kt

```diff
@@ -0,0 +1,196 @@
+import org.commcare.engine.resource.AndroidResourceUtils
+import org.commcare.engine.resource.ResourceInstallUtils
+import org.commcare.preferences.HiddenPreferences
+import org.commcare.resources.model.*
+import org.commcare.utils.AndroidCommCarePlatform
+import org.commcare.utils.AndroidUtil
+import org.commcare.utils.FileUtil
+import org.commcare.views.dialogs.PinnedNotificationWithProgress
+import org.javarosa.core.services.Logger
+import org.javarosa.core.util.SizeBoundUniqueVector
+import java.util.*
+import java.util.concurrent.TimeUnit
+import kotlin.collections.ArrayList
+
+// Contains helper functions to download lazy or missing media resources
+object MissingMediaDownloadHelper : TableStateListener, InstallCancelled {
+
+    var resourceInProgress: Resource? = null
+        private set
+
+    private val jobs = ArrayList<Job>()
+    private lateinit var mPinnedNotificationProgress: PinnedNotificationWithProgress
+
+
+    private const val BACK_OFF_DELAY = 5 * 60 * 1000L // 5 mins
+    private const val REQUEST_NAME = "missing_media_download_request"
+
+    var installCancelled: InstallCancelled? = null
+
+    // Schedules MissingMediaDownloadWorker
+    @JvmStatic
+    fun scheduleMissingMediaDownload() {
+        if (HiddenPreferences.shouldDownloadLazyMediaInBackground()) {
+            val constraints = Constraints.Builder()
+                    .setRequiredNetworkType(NetworkType.CONNECTED)
+                    .setRequiresBatteryNotLow(true)
+                    .build()
+
+            val downloadMissingMediaRequest = OneTimeWorkRequest.Builder(MissingMediaDownloadWorker::class.java)
+                    .addTag(CommCareApplication.instance().currentApp.appRecord.applicationId)
+                    .setConstraints(constraints)
+                    .setBackoffCriteria(
+                            BackoffPolicy.EXPONENTIAL,
+                            BACK_OFF_DELAY,
+                            TimeUnit.MILLISECONDS)
+                    .build()
+
+            WorkManager.getInstance(CommCareApplication.instance())
+                    .enqueueUniqueWork(
+                            getMissingMediaDownloadRequestName(),
+                            ExistingWorkPolicy.KEEP,
+                            downloadMissingMediaRequest)
+        }
+    }
+
+    // Returns Unique request name for the UpdateWorker Request
+    private fun getMissingMediaDownloadRequestName(): String {
+        val appId = CommCareApplication.instance().currentApp.uniqueId
+        return REQUEST_NAME + "_" + appId
+    }
+
+
+    // Verifies all application media and re-downloads any missing lazy resources
+    fun downloadAllMissingMedia() {
+        val platform = CommCareApplication.instance().commCarePlatform
+        val global = platform.globalResourceTable
+
+        global.setInstallCancellationChecker(this)
+
+        val problems = Vector<MissingMediaException>()
+        global.verifyInstallation(problems, platform)
```

Comment:
Have you timed this operation with CAS, out of curiosity?

Reply:
I have seen it varying from 14 to 30 seconds on lower mid range to mid range devices but can imagine it taking more closer to a minute on really low end devices.